src/laserscan_to_pointcloud_nodelet.cpp

```diff
@@ -0,0 +1,155 @@
+
+  // Check if explicitly single threaded, otherwise, let nodelet manager dictate thread pool size
+  if (concurrency_level == 1)
+  {
+    nh_ = getNodeHandle();
+  }
+  else
+  {
+    nh_ = getMTNodeHandle();
+  }
+
+  // Only queue one pointcloud per running thread
+  if (concurrency_level > 0)
+  {
+    input_queue_size_ = static_cast<unsigned int>(concurrency_level);
+  }
+  else
+  {
+    input_queue_size_ = boost::thread::hardware_concurrency();
+  }
+
+  // if pointcloud target frame specified, we need to filter by transform availability
+  if (!target_frame_.empty())
+  {
+    tf2_.reset(new tf2_ros::Buffer());
+    tf2_listener_.reset(new tf2_ros::TransformListener(*tf2_));
+    message_filter_.reset(new MessageFilter(sub_, *tf2_, target_frame_, input_queue_size_, nh_));
+    message_filter_->registerCallback(boost::bind(&LaserScanToPointCloudNodelet::scanCallback, this, _1));
+    message_filter_->registerFailureCallback(boost::bind(&LaserScanToPointCloudNodelet::failureCallback, this, _1, _2));
+  }
+  else  // otherwise setup direct subscription
+  {
+    sub_.registerCallback(boost::bind(&LaserScanToPointCloudNodelet::scanCallback, this, _1));
+  }
+
+  pub_ =
+      nh_.advertise<sensor_msgs::PointCloud2>("cloud", 10, boost::bind(&LaserScanToPointCloudNodelet::connectCb, this),
+                                              boost::bind(&LaserScanToPointCloudNodelet::disconnectCb, this));
+}
+
+void LaserScanToPointCloudNodelet::connectCb()
+{
+  boost::mutex::scoped_lock lock(connect_mutex_);
+  if (pub_.getNumSubscribers() > 0 && sub_.getSubscriber().getNumPublishers() == 0)
+  {
+    NODELET_INFO("Got a subscriber to cloud, starting laserscan subscriber");
+    sub_.subscribe(nh_, "scan_in", input_queue_size_);
+  }
+}
+
+void LaserScanToPointCloudNodelet::disconnectCb()
+{
+  boost::mutex::scoped_lock lock(connect_mutex_);
+  if (pub_.getNumSubscribers() == 0)
+  {
+    NODELET_INFO("No subscibers to cloud, shutting down subscriber to laserscan");
+    sub_.unsubscribe();
+  }
+}
+
+void LaserScanToPointCloudNodelet::failureCallback(const sensor_msgs::LaserScanConstPtr& scan_msg,
+                                                   tf2_ros::filter_failure_reasons::FilterFailureReason reason)
+{
+  NODELET_WARN_STREAM_THROTTLE(1.0, "Can't transform laserscan from frame " << scan_msg->header.frame_id << " to "
+                                                                            << message_filter_->getTargetFramesString()
+                                                                            << " at time " << scan_msg->header.stamp
+                                                                            << ", reason: " << reason);
+}
+
+void LaserScanToPointCloudNodelet::scanCallback(const sensor_msgs::LaserScanConstPtr& scan_msg)
+{
+  sensor_msgs::PointCloud2Ptr scan_cloud;
+  scan_cloud.reset(new sensor_msgs::PointCloud2);
+  projector_.projectLaser(*scan_msg, *scan_cloud);
```

Comment:
there's a more accurate way to transform a laserscan into a pointcloud, that takes into account the movement of the laser. it's probably not typically useful here if the target frame is moving (i.e. on a robot)

https://docs.ros.org/diamondback/api/laser_geometry/html/classlaser__geometry_1_1LaserProjection.html#ad5912a98fd3335a566256b28be688d25